v2/yarpcclient/provider.go

```diff
@@ -0,0 +1,50 @@
+// Copyright (c) 2018 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package yarpcclient
+
+import (
+	"go.uber.org/yarpc/v2"
+)
+
+var _ yarpc.ClientProvider = (*Provider)(nil)
+
+// Provider implements yarpc.ClientProvider.
+type Provider struct {
```

Comment:
I know this might be a little over-cautious, but we could add a `sync.RWMutex` here so that we can add read locking to `Client` and write locking to `Register`. It looks like this wasn't previously the case though, so not necessarily actionable right now.

Reply:
I don't expect there to be much read/write contention on the provider, but it's such a minimal change I'll add it.

Reply:
The provider must be immutable once lifecycles start. Adding a lock can only hurt. A note that register shouldn’t be called at run time would be better.

Reply:
fair point